Dev/improve preprocessing

a-benini-2 opened the pull request:
This pull request addresses some - not all - issues & goals brought up in issue _**Rework TBk preprocessing for modularization**_ #90:

- saving preprocessing outputs (raster) in an output (sub-)folder is enabled by two new advanced parameters `save_in_output_subfolder` & `output_subfolder` (f5dace26c59c9c43bdc83d9377e0ebcbc2b7978f):
<img width="726" height="622" alt="output_folder_2" src="https://github.com/user-attachments/assets/867ac5de-eed1-45ad-b9c8-1b48aee66340" />

At the moment `save_in_output_subfolder` is set to _False_ / unchecked by default, thus the preprocessing by default still saves the outputs in the familiar way / handing directory structure as previously. As the development of _**TBk**_ proceeds, we might switch this default-value to _True_.
- check if all preprocessing inputs have the same CRS (e5daa2a6bbf571a01dd0688a0c8e7118bacae5ef)
- enable virtual rasters (.vrt) as input for VHM and/or forest mixture degree (aa6de6abf694dccfbb485a493f586522b426ade0). See also issue #78.
- enable to clip VHM by the extent of the mask ([multi-]polygon-layer) as 1st preprocessing step by new advanced parameter `clip_vhm` (45b25baeaf2d91dc23fb0a06ad4f9fe6e5070b18).  
<img width="726" height="163" alt="clip_vhm" src="https://github.com/user-attachments/assets/cb738969-0ab7-42f2-a460-3a1437324bab" />

Note that clipping the input-VHM by the extent of the mask (default setting) lowers overall preprocessing run time, if extent of the mask is significantly smaller than extent of the VHM-input. If the VHM-input and the mask have about the same extent clipping may increase overall preprocessing run time. Implementing the clipping by mask under hood was not a straight forward processes as difficulties concerning the file type of the temp. clipped VHM, the data type of the input VHM and  run time performance had to be overcome  (4a184fee21155a1ea3117cfa6e71225d296b07f9, 60dfe5794ec0aa6dba3f568be781f47e0a6eca33, bf9a571123dacb0e5c26c5462a21c951bb4d140f, e3cece7a83a522bdb91696294019be814215ff78).

- enable to replace VHM NoData with zero by new advanced parameter `vNA_replacement` (c62b4cb3de0c27596afd5189649a781cd108b840, cff76d10f1c5a1b3b750ea52c34c0f79391f3d56)
<img width="837" height="277" alt="replace" src="https://github.com/user-attachments/assets/c2fe920e-c2bf-4dda-9f7f-dae05945b62f" />

- enable replacement of forest mixture degree NoData with two new advanced parameters `mg_NA_replacement` &` mg_NA_replacement_value` (f3f23df1bb94b1236aae3f2f8312154ed2c4b900, 45b8fd2ed1074f135380acc225a72e28b726d098, 4d0ac5f147a9081129fe2991bc0292690185a856)
<img width="727" height="294" alt="mg_replacement" src="https://github.com/user-attachments/assets/17e4e8c4-8576-4b9d-9ed6-a0247cfc0509" />

Note that if <i>Rescale Forest mixture values</i> is set to anything but 1 (no rescaling), <b><i>TBk prepare VHM (and MG)</i></b> replaces by default inevitably any NoData-pixels of <i>Forest Mixture Degree</i> with 0 (= 100% deciduous). If <i>Rescale Forest mixture values</i> is set to 1, NoData-pixels are preserved by default. By checking <i>Replacement of forest mixture degree NoData</i> and setting a numeric value as <i>Value for replacement</i> (default 0 = 100% deciduous) a non-default replacement of NoData-pixel is feasible, where the max. is 100 (= 100% coniferous). Setting <i>Value for replacement</i> either within the range of <i>Minimum</i> / <i>Maximum Deciduous</i> or of <i>Minimum</i> / <i>Maximum Coniferous</i> will convert the original NoData-pixels accordingly to 0 (= deciduous) resp. 100 (= coniferous) as pixel values of the <i>Binary mixture degree 10m output</i>.

Hannes-Ole committed:
feature: add function with graceful error handling for os.remove

also rearranged the utility functions

Hannes-Ole committed:
fix: add tmp file removal of nh_os rasters

Hannes-Ole committed:
fix: tbk classification rasters are now compressed

store_raster now uses compress configuration by default (DEFLATE, PREDICTOR = 2, ZLEVEL = 9)

Hannes-Ole committed:
fix: init coniferous mean in any case

This fixes the error "cannot access local variable 'coniferous_mean' where it is not associated with a value" that occurred mostly on recent MACs and a few Windows-PCs.

Hannes-Ole committed:
feat: BK Generate includes hdom_diff and local densities

Added the additional steps "calculate hdom difference" and "local densities" to the BK core stand generation (with parameter flags to turn these off). This is a convenience change since most projects require these steps anyway (this change will be made obsolete/redundant to the development in branch bk_algorithm_modularization, but is small enough to be worth it).

Hannes-Ole committed:
feat: TBk_Project_Template.qgz includes hdom_diff and local densities

As they are now part of the main workflow, styles/layers for "VHM-hdom difference" and "local densities" are included in the project template.
Also included themes and reworked some of the older themes.
TBk_QField_Themes.xlsx was updated to reflect new theme names. Obsolete additional tab was deleted.

Hannes-Ole committed:
feat: TBk_Project_Template.qgz includes hdom_diff and local densities

As they are now part of the main workflow, styles/layers for "VHM-hdom difference" and "local densities" are included in the project template.
Also included themes and reworked some of the older themes.
TBk_QField_Themes.xlsx was updated to reflect new theme names. Obsolete additional tab was deleted.

Hannes-Ole committed:
Merge remote-tracking branch 'origin/dev/main' into dev/main

# Conflicts:
#	tbk_qgis/tbk/bk_core/resources/TBk_Template.qgz

Hannes-Ole committed:
fix: TBk WIS2 export round all tree species values and fix sum == 0

WIS.2 can't handle floats and if the sum of tree species was 100, values were copied 1:1 and not rounded (resulting in float output). This is fixed now, by rounding when reading the values (guaranteeing integer output).

Another bug was fixed: If sum of tree species was 0, default behavior now is correctly setting p100 = 100.

Hannes-Ole committed:
fix: TBk WIS2 export works with feature sources

using parameterAsVectorLayer didn't work with using the "selected Features only" checkbox (and likely the .source() wouldn't work with temporary layers). Using parameterAsSource now works with both variants and seemingly has no negative side-effects.

Hannes-Ole committed:
feat: TBk WIS2 export default option to export DG per strata

introducing a checkbox, that is on by default (for the new WIS.2 web version)

Commit:
fix: update preprocessing test script

- algorithm "TBk:TBk prepare VHM (and MG)" replaces "TBk:TBk prepare VHM and MG"
- Grass7 is outdated
- import processing required

Commit:
feat: clip input VHM to extent of polygon mask

- as 1st step the detailed input VHM is clipped by an extent including the polygon mask as well as the extents of the outputs VHM 10m and VHM 150cm. This extent is aligned with the resolution of VHM input, which itself is assumed to be aligned with the CRS's origin (0, 0). VHM with fine pixel-resolution are seldom not aligned with the CRS's origin, exception may come in the form of "homegrown" VHM. So, under "normal" circumstances the extent generated for clipping should not result in any pixel-shifting of VHM input and thus neigher in any pixel-shifting of the output VHM_detail.tif. If the VHM input is not aligned to (0,0), a slight shifting of VHM_detail-pixels (< VHM-input-pixel-resolution) may take place (which we should test - at some time). Anyhow, alignment of VHM output (VHM 10m, VHM 150cm, VHM detail) of different TBk-projects, using the VHM-input, should be aligned with each other (this needs also to be tested).
- the clipped VHM is saved by default as a temporary virtual raster (tmp_vhm_clipped.tif), which is then feed into the further VHM processing. Keeping or deleting tmp_vhm_clipped.tif is controlled by the advanced parameter "Delete temporary files".
- Clipping the VHM input reduces the data-load, if the extent of the mask polygon ist smaller than the one of the original VHM. Thus, this should speed up the preprocessing. vrt.files are lightweight.

Commit:
fix: delete VHM_10m.tif.aux.xml & VHM_150cm.tif.aux.xml

if they happen to exist in directory, where preprocessing outputs are dropped. Analog to treatment of VHM_detail.tif.aux.xml.

Commit:
fix: fix typos

Commit:
fix: clean up code concerning output directory

- eliminate working_root, which was a synonym of output_root
- drop unregistered parameter/output OUTPUT, use registered OUTPUT_ROOT instead

Commit:
fix: make parameter MG_RESCALE_FACTOR usable

- correct parameter key
  MG_RESCALE_FACTOR = "100" --> MG_RESCALE_FACTOR = "mg_rescale_factor"
- tests with several MG inputs having different min/max ranges show that now the advanced parameter "Rescale Forest mixture degree values ..." is actually working as intended. Prior, processing only worked with MG inputs having a min/max range [0, 10^4].

Commit:
fix: update preprocessing test script

... according to fix of preprocessing algorithm

Commit:
feat: allow virtual rasters (.vrt) as input for VHM and/or forest mixture degree

Commit:
feat: check if CRSs of inputs are identical

else stop and throw an according error message

Commit:
feat: enable overwriting NoData-values of VHM with an integer number

- by default not active, to activate checkbox "Replacement of VHM NoData" required
- if the VHM is masked, replacement of NoData-values only affects the area covered by the polygon mask
- the integer value used for replacement must be >= 0 and =< VHM NoData value
- the replacement value is by default 0 --> assuming that NoData = no (woody) vegetation
- if the replacement value is set to something > 0, like 99 or 199, this might be of interest to check out potential flaws of the input VHM, but such TBk-preprocessing VHM-outputs are not appropriate to use as inputs for Generate BK.

Commit:
feat: update preprocessing test script

include new advanced parameters:
- vNA_replacement
- vNA_replacement_value

Commit:
refactor: prefix all temp. files related to VHM with vhm_

Commit:
feat: enable overwriting NoData-values of forest mixture degree input by integer value

- by default not active, to activate checkbox "Replacement of forest mixture degree NoData" required
- the integer value (percentage) used for replacement must be >= 0% (only deciduous) and =< 100% (only coniferous)
- the replacement value is by default 0 --> assuming that NoData = 100% coverage by deciduous trees
- at the time TBk-preprocessing inevitably turns any NoData-pixel of the forest mixture input by default into 0 (= 100% deciduous), this means that NoData-values are not preserved (which seems to be hard to fix). So the advanced parameter  "Replacement of forest mixture degree NoData" and "Value for replacement of ... " are sort of a qick fix to enable non-default handling of NoData-values of forest mixture degree input. So far forest mixture degree rasters with NoData pixel overlapping with area covered by the polygon mask at hand have been rather exceptional.

Commit:
feat: update preprocessing test script

include new advanced parameters:
- mg_NA_replacement
- mg_NA_replacement_value

Commit:
fix: fix typo

Commit:
fix: ensure data type of MG_10m & MG_10m_binary is always byte

- prior if no rescaling was applied to MG (rescal factor = 1) MG_10m be would simply a copy the aligned MG, thus inheriting the data type of the MG input.
- if "Replacement of forest mixture degree NoData" is NOT applied:
  -- > this commit leaves the handling of NoData-pixels included in MG input uneffected as is: In case of rescaling MG's NoData pixels are returned as pixels with value = 0 in MG_10m & MG_10m_binary. If no rescaling takes place, NoData pixels are preserved.
- if "Replacement of forest mixture degree NoData" is applied:
   -- > this commit now enables applying this advanced parameter in combination with no rescaling (rescal factor = 1). Prior, combining these advance parameter settings would have resulted in an error / a crash of TBk-preprocessing.

Commit:
fix: correct description

according to prior commit

Commit:
fix: simplify handling of VHM's NoData-pixels

- change advanced parameter "Replacement of VHM NoData" (checkbox) to "Replacment of VHM NoData with 0"
- thus drop advanced parameter "Value for replacement of VHM NoData"

Commit:
fix: fix update of preprocessing test script

Commit:
feat: align MG input as .vrt instead of .tif

- should reduce the volume of temp. data during TBk-preprocessing

Commit:
fix: save clipped input-VHM as .tif

Saving the VHM-input clipped by the extent of the forest mask as virtual raster layer (.vrt) is as such a fast process. But compared to a .tif the .vrt-format slows down the following preprocessing. This is very significant when clipping from larger VHM-inputs. Thus, saving the clipped VHM-input is switched from .vrt to .tif and thereby lowering overall preprocessing run time.

Commit:
fix: enable deleting temp. clipped VHM-input

the trick: applying gdal.Open() to the original / unclipped VHM-input instead of to the clipped VHM.

Commit:
fix: enable to safely delete temp. aligned MG (.vrt)

Commit:
feat: enable to leave out clipping VHM-input

clipping the input-VHM by the extent of the mask (default setting) lowers overall preprocessing run time, if extent of the mask is significantly smaller than extent of the VHM-input. If the VHM-input and the mask have about the same extent clipping may increase overall preprocessing run time.

Commit:
feat: update preprocessing test script

... according to latest feat of preprocessing algorithm

Commit:
refactor: get data type of input-VHM only once

Commit:
feat: enable saving preprocessing outputs in subfolder

... within output folder. 2 new advanced parameters make this possible

1) save_in_output_subfolder, boolean / checkbox. At the moment set to False by default, thus the preprocessing by default still saves the outputs in the familiar way / handing directory structure as previously. As the development of TBk proceeds, we might switch this default-value to True.
2) output_subfolder string / name of subfolder: default base_data_preprocessed.

Commit:
feat: update preprocessing test script

... according to latest feat of preprocessing algorithm

a-benini-2 commented:
I have tested if the improved version of _**TBk prepare VHM (and MG)**_ is capable to get its inputs from HAFL's P-drive and drop  its outputs on the P-drive. And it did work both with my laptop via the internet and with HARA04. During these tests I've set the new advanced parameter `clip_vhm` to _True_ and thus was able to clip the raster-inputs VHM & forest mixture degree covering the whole Swiss territory with extents of two forest masks of smaller Cantons  (BS, AI). 

My provisional conclusion: _**TBk prepare VHM (and MG)**_ is able to run with network paths. Thus, issue #87 seems to be resolved and the goal _Network compatibility_ brought up in issue #90 seems to be also achieved. Can someone confirm the network compatibility of _**TBk prepare VHM (and MG)**_ ? We should test with larger (forest-)masks and not just only with the P-drive.

a-benini-2 commented:
There's news about _**TBk prepare VHM (and MG)**_ 's network compatibility:

1) Specifying the _**Output folder**_ with 
`\\bfh.ch\data\HAFL\7 WWI\74a FF WG\742a Aktuell\L.012359-52-WFOM_TBk_II\00_TBk_Kleinprojekte\2025-10_FBBueren\Daten\base_data`
does not work!
Note that advanced parameters _**save_in_output_subfolder**_ and _**output_subfolder**_ were set to `True` resp. `base_data_preprocessed_2023`.
<img width="1037" height="319" alt="setting" src="https://github.com/user-attachments/assets/e56a3e35-ce1f-4add-9fe7-1efb31b5bd9d" />
<img width="630" height="547" alt="errror" src="https://github.com/user-attachments/assets/74abf9ef-f2cc-470e-b295-f59d61f6f096" />

Relevant error message: 
`Unable to execute algorithm
Could not load source layer for INPUT: \\bfh.ch\data\HAFL\7 WWI\74a FF WG\742a Aktuell\L.012359-52-WFOM_TBk_II\00_TBk_Kleinprojekte\2025-10_FBBueren\Daten\base_data\base_data_preprocessed_2023\vhm_clipped.tif not found`

2) Specifying the _**Output folder**_ with 
`P:\HAFL\7 WWI\74a FF WG\742a Aktuell\L.012359-52-WFOM_TBk_II\00_TBk_Kleinprojekte\2025-10_FBBueren\Daten\base_data`
works perfectly fine (run time on HARA04 ≈ 2.5 minutes)!

Conclusion:
While paths to input files starting with `\\bfh.ch\data\HAFL\...`   do work for _**TBk prepare VHM (and MG)**_ , starting the _**Output folder**_-path with `\\bfh.ch\data\HAFL\...` does not! Starting the _**Output folder**_-path with `P:\HAFL\...` seems to ensure network compatibility - at least for  _**TBk prepare VHM (and MG)**_. We have to investigate, if this is also the solution for network compatibility issues of other _**TBk**_ algorithms. Possibly this is also linked to difficulties when it come to deleting temp. vector layers / .gpkg.

Hannes-Ole commented:
Nice work. I tested this on TBk Moutier and it worked like a charm.

I ran into two problems though, that would be nice QoL improvements:
- [ ] I tried using "swissBOUNDARIES3D_1_5_LV95_LN02 — tlm_hoheitsgebiet" -> Selected Features only, which resulted in a "NoneType" error (i.e. apparently it isn't possible to use the "selected features only"-functionality? (see input parameters below; I noticed that a "/bfh.ch/data" sneaked itself in there, maybe that is a result of the "Selected Features only" part.
- [ ] I then tried creating a quick temporary layer (simply doing a quick buffer) to use that. There I got the error that "VHM input and polygon mask must have the same CRS" - which they did. Persisting the temporary layer solved the problem.

Another thing for consistency reasons in terms of design / can save two parameters:
- [ ] I'd prefer the default design to avoid boolean parameters followed by a parameter to specify as is the case with Replacement of forest mixture degree NoData (Boolean) and Value for replacement (Real/Int). Rather use a single parameter (Real/Int); if unspecified, it defaults to a the Boolean = False (the parameter isn't used), if specified Boolean = True (the parameter is used). 
- [ ] same design decision applies to the "Save preprocessing outputs in subfolder" (Boolean) + "Name of subfolder" (String)

<img width="534" height="91" alt="image" src="https://github.com/user-attachments/assets/21697832-9bfb-46ea-a47a-489209b312fc" />


______________________

Input for using "selected features only" as the polygon mask

```
processing.run("TBk:TBk prepare VHM (and MG)", {'vhm_input':'//bfh.ch/data/HAFL/9 Share/PermanenteOrdner/Geodaten/Nationale_Daten/Nationale_Daten_SWISSTOPO/SwissSurface3D_VHM/2025-02_VHM_swissSurface3D/20250217_VHM_ALS_CH_SWISS_1m_2056_BYTE.tif','mg_input':'//bfh.ch/data/HAFL/9 Share/PermanenteOrdner/Geodaten/Nationale_Daten/Nationale_Daten_WSL/Waldmischungsgrad_LFI_2023/wmg2024_onerf_s1s2as_nomask_lv95neu-tfw_2025-03-04_1610/WMG2025_oneRF_S1S2AS_nomask_LV95_NH_INT16.tif','mask':QgsProcessingFeatureSourceDefinition('//bfh.ch/data/HAFL/9 Share/PermanenteOrdner/Geodaten/Nationale_Daten/Nationale_Daten_SWISSTOPO/SwissBoundaries_2025/swissBOUNDARIES3D_1_5_LV95_LN02.gpkg|layername=tlm_hoheitsgebiet', selectedFeaturesOnly=True, featureLimit=-1, geometryCheck=QgsFeatureRequest.GeometryAbortOnInvalid),'output_root':'P:\\HAFL\\7 WWI\\74a FF WG\\742a Aktuell\\L.012359-52-WFOM_TBk_II\\00_TBk_Kleinprojekte\\2025-11_Moutier\\Daten','save_in_output_subfolder':True,'output_subfolder':'base_data_preprocessed','vhm_detail':'VHM_detail.tif','vhm_10m':'VHM_10m.tif','vhm_150cm':'VHM_150cm.tif','mg_10m':'MG_10m.tif','mg_10m_binary':'MG_10m_binary.tif','align_method':0,'del_tmp':True,'clip_vhm':True,'mask_vhm':True,'vhm_convert_to_byte':True,'vhm_reclassify':False,'vMin':0,'vMax':60,'vNA':255,'vNA_replacement':False,'mg_rescale_factor':100,'reclassify_mg_values':True,'min_lh':0,'max_lh':50,'min_nh':50,'max_nh':100,'mg_NA_replacement':False,'mg_NA_replacement_value':0})

Input parameters:
{ 'align_method' : 0, 'clip_vhm' : True, 'del_tmp' : True, 'mask' : QgsProcessingFeatureSourceDefinition('//bfh.ch/data/HAFL/9 Share/PermanenteOrdner/Geodaten/Nationale_Daten/Nationale_Daten_SWISSTOPO/SwissBoundaries_2025/swissBOUNDARIES3D_1_5_LV95_LN02.gpkg|layername=tlm_hoheitsgebiet', selectedFeaturesOnly=True, featureLimit=-1, geometryCheck=QgsFeatureRequest.GeometryAbortOnInvalid), 'mask_vhm' : True, 'max_lh' : 50, 'max_nh' : 100, 'mg_10m' : 'MG_10m.tif', 'mg_10m_binary' : 'MG_10m_binary.tif', 'mg_NA_replacement' : False, 'mg_NA_replacement_value' : 0, 'mg_input' : '//bfh.ch/data/HAFL/9 Share/PermanenteOrdner/Geodaten/Nationale_Daten/Nationale_Daten_WSL/Waldmischungsgrad_LFI_2023/wmg2024_onerf_s1s2as_nomask_lv95neu-tfw_2025-03-04_1610/WMG2025_oneRF_S1S2AS_nomask_LV95_NH_INT16.tif', 'mg_rescale_factor' : 100, 'min_lh' : 0, 'min_nh' : 50, 'output_root' : 'P:\\HAFL\\7 WWI\\74a FF WG\\742a Aktuell\\L.012359-52-WFOM_TBk_II\\00_TBk_Kleinprojekte\\2025-11_Moutier\\Daten', 'output_subfolder' : 'base_data_preprocessed', 'reclassify_mg_values' : True, 'save_in_output_subfolder' : True, 'vMax' : 60, 'vMin' : 0, 'vNA' : 255, 'vNA_replacement' : False, 'vhm_10m' : 'VHM_10m.tif', 'vhm_150cm' : 'VHM_150cm.tif', 'vhm_convert_to_byte' : True, 'vhm_detail' : 'VHM_detail.tif', 'vhm_input' : '//bfh.ch/data/HAFL/9 Share/PermanenteOrdner/Geodaten/Nationale_Daten/Nationale_Daten_SWISSTOPO/SwissSurface3D_VHM/2025-02_VHM_swissSurface3D/20250217_VHM_ALS_CH_SWISS_1m_2056_BYTE.tif', 'vhm_reclassify' : False }
```

```
Traceback (most recent call last):
File "D:\Users/hbh1/AppData/Roaming/QGIS/QGIS3\profiles\default/python/plugins\tbk_qgis\tbk\preproc\tbk_qgis_prepare_vhm_mg_algorithm.py", line 411, in processAlgorithm
mask = str(self.parameterAsVectorLayer(parameters, self.MASK, context).source())
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'source'
```

<img width="562" height="546" alt="image" src="https://github.com/user-attachments/assets/8692cfa5-4844-4b03-a186-cbee4920d882" />

a-benini-2 commented:
When I tried to run _**Generate BK**_ on HARA04  with a selected perimeter input, I received similar error message:
<img width="481" height="547" alt="generate_bk_selection_error" src="https://github.com/user-attachments/assets/a54a2c38-1f03-41a1-a4fe-ff8087e4e8c9" />
I conclude that the inability to use selected vector input is not restricted to _**TBk perpare VHM (and MG)**_, but something more common among _**TBk**_-algorithm.

After saving the selected geometries form _swissBOUNDARIES3D_1_5_LV95_LN02 — tlm_hoheitsgebiet_ as a .gpkg and then using this without any buffering as mask-input for _**TBk perpare VHM (and MG)**_ worked on HARA04 / P-drive perfectly fine.

Overcoming the inability to use selected vector input might be an issue to be solved by OPENGIS.

Commit:
fix: simplify replacing MG NoData

- drop advanced parameter mg_NA_replacement
- advanced parameter mg_NA_replacement_value becomes optional / not set by default

Commit:
fix: update preprocessing test script

... according to latest fix of preprocessing algorithm

a-benini-2 commented:
- I've added a commit simplifying the (non-default) replacement of forest mixture degree NoData. https://github.com/HAFL-WWI/TBk/pull/95/commits/607e823fe54e8d3671cd09427264d7f87082ee74

- I did not simplify the handling of the optional subfolder to gather preprocessing outputs within the output folder. It is possible to set the name of the subfolder by default to None resp. making this advanced parameter optional. Thus there would be no need  for a checkbox _Save preprocessing outputs in subfolder_. But without checkbox it's not possible to to set a specific string as default name for the subfolder and making it a the same time optional by entering None / nothing, which should result in dropping the preprocessing outputs directly into the output folder. Working without checkbox _Save preprocessing outputs in subfolder_ and having a default subfolder name, allows to use an alternative subfolder name, but when specified as None, the outputs would be still gathered in the default named subfolder within the output folder.

a-benini-2 committed:
Merge branch 'HAFL-WWI:main' into dev/improve-preprocessing
